[No QA] [TS Migration] Adjust guidelines, both TS and JS markdown files (#41212)
Expensify/App: contributingGuides/STYLE.md

```ts
// BAD
const foo: string = "foo";
const [counter, setCounter] = useState<number>(0);

// GOOD
const foo = "foo";
const [counter, setCounter] = useState(0);
const [username, setUsername] = useState<string | undefined>(undefined); // Username is a union type of string and undefined, and its type cannot be inferred from the default value of undefined
```

For function return types, default to always typing them unless a function is simple enough to reason about its return type.

> Why? Explicit return type helps catch errors when implementation of the function changes. It also makes it easy to read code even when TypeScript intellisense is not provided.

```ts
function simpleFunction(name: string) {
    return `hello, ${name}`;
}

function complicatedFunction(name: string): boolean {
// ... some complex logic here ...
    return foo;
}
```

### Utility Types

Use types from [TypeScript utility types](https://www.typescriptlang.org/docs/handbook/utility-types.html) and [`type-fest`](https://github.com/sindresorhus/type-fest) when possible.

```ts
type Foo = {
    bar: string;
};

// BAD
type ReadOnlyFoo = {
    readonly [Property in keyof Foo]: Foo[Property];
};

// GOOD

> type ReadOnlyFoo = Readonly<Foo>;

Reviewer: Can we please also add an example for `ValueOf`?

```ts
// BAD
type FooValue = Foo[keyof Foo];

// GOOD
type FooValue = ValueOf<Foo>;
```

```

### `object` type

Don't use `object` type. eslint: [`@typescript-eslint/ban-types`](https://typescript-eslint.io/rules/ban-types/)

> Why? `object` refers to "any non-primitive type," not "any object". Typing "any non-primitive value" is not commonly needed.

```ts
// BAD
const foo: object = [1, 2, 3]; // TypeScript does not error
```

If you know that the type of data is an object but don't know what properties or values it has beforehand, use `Record<string, unknown>`.

> Even though `string` is specified as a key, `Record<string, unknown>` type can still accepts objects whose keys are numbers. This is because numbers are converted to strings when used as an object index. Note that you cannot use [symbols](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) for `Record<string, unknown>`.

```ts
function logObject(object: Record<string, unknown>) {
    for (const [key, value] of Object.entries(object)) {
        console.log(`${key}: ${value}`);
    }
}
```

### Prop Types

> Don't use `ComponentProps` to grab a component's prop types. Go to the source file for the component and export prop types from there. Import and use the exported prop types.

Reviewer: I'm actually not sure why we decided this, so maybe we should add a `why` clause here?

Author: I'm not 100% sure either, as it was copied from `TS_STYLE.md`. I think it might be mostly for consistency, I have also seen some threads with people complaining about `ComponentProps` not always working, e.g. for class components with default props. Any ideas what we should put in the **why** section? 

CC: @fabioh8010 @blazejkustra

Reviewer 2: Yes, since we have our own props defined in the codebase is always easy to just export and use it when required. `ComponentProps` (and its friends `ComponentPropsWithoutRef`, ComponentPropsWithRef`) should be used only for specific situations or generic components.

Author: Thanks guys, changes applied!

> Don't export prop types from component files by default. Only export it when there is a code that needs to access the prop type directly.

```tsx
// MyComponent.tsx
export type MyComponentProps = {
    foo: string;
};

export default function MyComponent({ foo }: MyComponentProps) {
    return <Text>{foo}</Text>;
}

// BAD
import { ComponentProps } from "React";
import MyComponent from "./MyComponent";
type MyComponentProps = ComponentProps<typeof MyComponent>;

// GOOD
import MyComponent, { MyComponentProps } from "./MyComponent";
```

### File organization

In modules with platform-specific implementations, create `types.ts` to define shared types. Import and use shared types in each platform specific files. Do not use [`satisfies` operator](#satisfies-operator) for platform-specific implementations, always define shared types that complies with all variants.

> > Why? To encourage consistent API across platform-specific implementations. If you're migrating module that doesn't have a default implement (i.e. `index.ts`, e.g. `getPlatform`), refer to [Migration Guidelines](#migration-guidelines) for further information.

Reviewer: ```suggestion
> Why? To encourage consistent API across platform-specific implementations. If you're migrating module that doesn't have a default implementation (i.e. `index.ts`, e.g. `getPlatform`), refer to [Migration Guidelines](#migration-guidelines) for further information.
```

> Why? Prop types might [need to be exported from component files](#export-prop-types). If the component is only used inside a file or module and not exported, then inline prop types can be used.

```ts
// BAD
export default function MyComponent({ foo, bar }: { foo: string, bar: number }){
   // component implementation
};

// GOOD
type MyComponentProps = { foo: string, bar: number };
export default MyComponent({ foo, bar }: MyComponentProps){
   // component implementation
}
```

### Satisfies Operator

Use the `satisfies` operator when you need to validate that the structure of an expression matches a specific type, without affecting the resulting type of the expression.

> Why? TypeScript developers often want to ensure that an expression aligns with a certain type, but they also want to retain the most specific type possible for inference purposes. The `satisfies` operator assists in doing both.

```ts
// BAD
const sizingStyles = {
    w50: {
        width: '50%',
    },
    mw100: {
        maxWidth: '100%',
    },
} as const;

// GOOD
const sizingStyles = {
    w50: {
        width: '50%',
    },
    mw100: {
        maxWidth: '100%',
    },

> } satisfies Record<string, ViewStyle>;

Reviewer: *question:* Would you need to do this to get the most narrow type?

```suggestion
} as const satisfies Record<string, ViewStyle>;
```

Author: I think so, but @fabioh8010 @blazejkustra can you confirm?

Reviewer 2: > _question:_ Would you need to do this to get the most narrow type?

Yes that is correct, we could put another example as there are cases `as const satisfies` is not desirable.

Reviewer 3: I think it makes a difference in some cases, example:

satisfies Record<string, ViewStyle>:
```ts
    br0: {
        borderRadius: number;
    };
    br1: {
        borderRadius: number;
    };
}
```
as const satisfies Record<string, ViewStyle>:
```ts
{
    readonly br0: {
        readonly borderRadius: 0;
    };
    readonly br1: {
        readonly borderRadius: 4;
    };
}
```

So if we see any value in such narrow type I'm good to adjust the guidelines 👍

Imports:
```ts
// BAD
import {SomeType} from './a'
import someVariable from './a'

import {someVariable, SomeOtherType} from './b'

// GOOD
import type {SomeType} from './a'
import someVariable from './a'
```

 Exports:
```ts
// BAD
export {SomeType}
export someVariable
// or 
export {someVariable, SomeOtherType}

// GOOD
export type {SomeType}
export someVariable
```

### Exception to Rules

Most of the rules are enforced in ESLint or checked by TypeScript. If you think your particular situation warrants an exception, post the context in the `#expensify-open-source` Slack channel with your message prefixed with `TS EXCEPTION:`. The internal engineer assigned to the PR should be the one that approves each exception, however all discussion regarding granting exceptions should happen in the public channel instead of the GitHub PR page so that the TS migration team can access them easily.

When an exception is granted, link the relevant Slack conversation in your PR. Suppress ESLint or TypeScript warnings/errors with comments if necessary.

> This rule will apply until the migration is done. After the migration, discussion on granting exception can happen inside the PR page and doesn't need take place in the Slack channel.

Reviewer 3: I think we could remove this paragraph, as this is a dead rule that nobody applied during migration 🤷‍♂️

### Communication Items

> Comment in the `#expensify-open-source` Slack channel if any of the following situations are encountered. Each comment should be prefixed with `TS ATTENTION:`. Internal engineers will access each situation and prescribe solutions to each case. Internal engineers should refer to general solutions to each situation that follows each list item.

- I think types definitions in a third party library or JavaScript's built-in module are incomplete or incorrect

When the library indeed contains incorrect or missing type definitions and it cannot be updated, use module augmentation to correct them. All module augmentation code should be contained in `/src/types/modules/*.d.ts`, each library as a separate file.

```ts
// external-library-name.d.ts

declare module "external-library-name" {
    interface LibraryComponentProps {
        // Add or modify typings
        additionalProp: string;
    }
}

> ```

Reviewer 3: Same here, I don't think we need it. The only message with "TS ATTENTION" from slack was form Hayata [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1691772060882829) 😄

## Naming Conventions

### Type names

  - Use PascalCase for type names. eslint: [`@typescript-eslint/naming-convention`](https://typescript-eslint.io/rules/naming-convention/)

    ```ts
    // BAD
    type foo = ...;
    type BAR = ...;

    // GOOD
    type Foo = ...;
    type Bar = ...;
    ```

  - Do not postfix type aliases with `Type`.

    ```ts
    // BAD
    type PersonType = ...;

    // GOOD
    type Person = ...;
    ```

  - Use singular name for union types.

    ```ts
    // BAD
    type Colors = "red" | "blue" | "green";

    // GOOD
    type Color = "red" | "blue" | "green";
    ```

  - Use `{ComponentName}Props` pattern for prop types.

    ```ts
    // BAD
    type Props = {
        // component's props
    };

    function MyComponent({}: Props) {
        // component's code
    }

    // GOOD
    type MyComponentProps = {
        // component's props
    };

    function MyComponent({}: MyComponentProps) {
        // component's code
    }
    ```

  - Use {ComponentName}Handle for custon ref handle types.

    ```tsx
    // BAD
    type MyComponentRef = {
        onPressed: () => void;
    };

    // GOOD

>     type MyComponentHandle = {

Reviewer 3: We may need to remove this rule or adjust the code as the rule isn't really used
![image](https://github.com/Expensify/App/assets/46095609/033336d3-4764-4260-b93c-a9b4f9d32f6a)
vs
![image](https://github.com/Expensify/App/assets/46095609/31e6806b-268f-4e5f-bdcc-d838c9c4056d)

> Using named functions is the preferred way to write a callback method.

Reviewer 3: I don't think this rule is followed in the codebase, I think we could remove it.

```

> Empty functions (noop) should be declared as arrow functions with no whitespace inside. Avoid _.noop()

Reviewer 3: This rule is useless imo, handled automatically with eslint + we no longer use lodash and underscore so we don't have to worry about _.noop usage

## Object / Array Methods

> We have standardized on using the native [Array instance methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array#instance_methods) instead of [underscore.js](https://underscorejs.org/) methods for objects and collections. As the vast majority of code is written in TypeScript, we can safaly use the native methods.

Reviewer 3: ```suggestion
We have standardized on using the native [Array instance methods](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array#instance_methods) instead of [lodash](https://lodash.com/) methods for objects and collections. As the vast majority of code is written in TypeScript, we can safely use the native methods.
```

Here are a couple of things we would ask that you *avoid* to help maintain consistency in our codebase:

- **Async/Await** - Use the native `Promise` instead

> - **Optional Chaining** - Yes, don't use `lodashGet()`
> - **Null Coalescing Operator** - Yes, don't use `lodashGet()` or `||` to set a default value for a possibly `undefined` or `null` variable

Reviewer 3: ```suggestion
```

Reviewer 3: This doesn't make sense with this header: "Here are a couple of things we would ask that you *avoid* to help maintain consistency in our codebase:"

    ...
    return <input {...props} ref={ref} />
};

export default React.forwardRef(FancyInput)
```

If the ref handle is not available (e.g. `useImperativeHandle` is used) you can define a custom handle type above the component.

```tsx 
import type {ForwarderRef} from 'react';
import {useImperativeHandle} from 'react';

type FancyInputProps = {
    ...
    onButtonPressed: () => void;
};

type FancyInputHandle = {
  onButtonPressed: () => void;
}

function FancyInput(props: FancyInputProps, ref: ForwardedRef<FancyInputHandle>) {
    useImperativeHandle(ref, () => ({onButtonPressed}));

    ...
    return <input {...props} ref={ref} />;
};

export default React.forwardRef(FancyInput)
```

### Hooks and HOCs

Use hooks whenever possible, avoid using HOCs.

> Why? Hooks are easier to use (can be used inside the function component), and don't need nesting or `compose` when exporting the component. It also allows us to remove `compose` completely in some components since it has been bringing up some issues with TypeScript. Read the [`compose` usage](#compose-usage) section for further information about the TypeScript issues with `compose`.

> > Note: Because Onyx doesn't provide a hook yet, in a component that accesses Onyx data with `withOnyx` HOC, please make sure that you don't use other HOCs (if applicable) to avoid HOC nesting.

Reviewer: This is out of date - let's update this to formalize  `useOnyx` as a best practice over `withOnyx`

```tsx
// BAD
type ComponentOnyxProps = {
    session: OnyxEntry<Session>;
};

type ComponentProps = WindowDimensionsProps &
    WithLocalizeProps &
    ComponentOnyxProps & {
        someProp: string;
    };

function Component({windowWidth, windowHeight, translate, session, someProp}: ComponentProps) {
    // component's code
}

export default compose(
    withWindowDimensions,
    withLocalize,
    withOnyx<ComponentProps, ComponentOnyxProps>({
        session: {
            key: ONYXKEYS.SESSION,
        },
    }),
)(Component);

// GOOD
type ComponentOnyxProps = {
    session: OnyxEntry<Session>;
};

type ComponentProps = ComponentOnyxProps & {
    someProp: string;
};

function Component({session, someProp}: ComponentProps) {
    const {windowWidth, windowHeight} = useWindowDimensions();
    const {translate} = useLocalize();
    // component's code
}

// There is no hook alternative for withOnyx yet.

> export default withOnyx<ComponentProps, ComponentOnyxProps>({

Reviewer: Let's update this for `useOnyx` too

    session: {
        key: ONYXKEYS.SESSION,
    },
})(Component);
```

### Stateless components vs Pure Components vs Class based components vs Render Props - When to use what?

Class components are DEPRECATED. Use function components and React hooks.

[https://react.dev/reference/react/Component#migrating-a-component-with-lifecycle-methods-from-a-class-to-a-function](https://react.dev/reference/react/Component#migrating-a-component-with-lifecycle-methods-from-a-class-to-a-function)

### Composition 

Avoid the usage of `compose` function to compose HOCs in TypeScript files. Use nesting instead.

>   > Why? `compose` function doesn't work well with TypeScript when dealing with several HOCs being used in a component, many times resulting in wrong types and errors. Instead, nesting can be used to allow a seamless use of multiple HOCs and result in a correct return type of the compoment. Also, you can use [hooks instead of HOCs](#hooks-instead-of-hocs) whenever possible to minimize or even remove the need of HOCs in the component.

Reviewer: ```suggestion
> Why? `compose` function doesn't work well with TypeScript when dealing with several HOCs being used in a component, many times resulting in wrong types and errors. Instead, nesting can be used to allow a seamless use of multiple HOCs and result in a correct return type of the compoment. Also, you can use [hooks instead of HOCs](#hooks-instead-of-hocs) whenever possible to minimize or even remove the need of HOCs in the component.
```
